include/pybind11/detail/class_.h

```diff
@@ -0,0 +1,427 @@
+/*
+    pybind11/detail/class_.h: Metatype, static properties, base types
+
+    Copyright (c) 2017 Wenzel Jakob <wenzel.jakob@epfl.ch>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "../attr.h"
+
+NAMESPACE_BEGIN(pybind11)
+NAMESPACE_BEGIN(detail)
+
+#if !defined(PYPY_VERSION)
+
+extern "C" inline PyObject *static_descr_get(PyObject *self, PyObject * /*obj*/, PyObject *type) {
```

Comment:
Ok -- I think it is fairly readable even with the separation. In case you're still willing to make a few changes, then one thing that would be nice to have is a 1-line comment for each hook that says roughly why its' needed. E.g.

```cpp
/// Traverse the internal dictionary of an instance. Used to implement garbage collection on types with the dynamic_attr tag.
extern "C" inline int object_traverse(PyObject *self, visitproc visit, void *arg) {
```
(This one is especially confusing since the surrounding context is missing in the non-lambda implementation.)